documentation/asciidoc/computers/compute-module/cm-bootloader.adoc

```diff
@@ -1,41 +1,55 @@
 
-* Selecting and verifying a specific bootloader release. The version in the `usbboot` repo is always a recent stable release.
-* Configuring the boot device (e.g. network boot). See `BOOT_ORDER` section in the xref:raspberry-pi.adoc#raspberry-pi-bootloader-configuration[bootloader configuration] guide.
-* Enabling hardware write-protection on the bootloader EEPROM to ensure that the bootloader can't be modified on inaccessible products (such as remote or embedded devices).
+* Select a specific bootloader release. Verify that every Compute Module you use has that release. The version in the `usbboot` repo is always a recent stable release.
+* Configure the boot device by xref:raspberry-pi.adoc#raspberry-pi-bootloader-configuration[setting the `BOOT_ORDER` ].
+* Enable hardware write-protection on the bootloader EEPROM to ensure that the bootloader can't be modified on inaccessible products (such as remote or embedded devices).
 
-NOTE: The Compute Module 4 ROM never runs `recovery.bin` from SD/eMMC and the `rpi-eeprom-update` service is not enabled by default. This is necessary because the eMMC is not removable and an invalid `recovery.bin` file would prevent the system from booting. This can be overridden and used with `self-update` mode where the bootloader can be updated from USB MSD or network boot. However, `self-update` mode is not an atomic update and therefore is not safe in the event of a power failure while the EEPROM is being updated.
+=== Flash Compute Module bootloader EEPROM
 
-==== Flash storage devices other than SD cards
+To flash the bootloader EEPROM:
 
-The Linux-based https://github.com/raspberrypi/usbboot/blob/master/mass-storage-gadget/README.md[mass-storage gadget] supports flashing of NVMe, eMMC and USB block devices. This is normally faster than using the `rpiboot` firmware driver and also provides a UART console to the device for easier debug.
+. Set up the hardware as you would when xref:../computers/compute-module.adoc#flash-compute-module-emmc[flashing the eMMC], but ensure `EEPROM_nWP` is _not_ pulled low.
+. Once complete, `EEPROM_nWP` may be pulled low again.
+. Run the following command to write `recovery/pieeprom.bin` to the bootloader EEPROM:
++
+[source,console]
+----
+$ ./rpiboot -d recovery
+----
 
-See also: https://github.com/raspberrypi/usbboot/blob/master/Readme.md#compute-module-4-extensions[CM4 rpiboot extensions].
+=== Flash storage devices other than SD cards
 
-==== Modify the bootloader configuration
+The Linux-based https://github.com/raspberrypi/usbboot/blob/master/mass-storage-gadget/README.md[`mass-storage-gadget`] supports flashing of NVMe, eMMC and USB block devices. `mass-storage-gadget` works faster than `rpiboot` and also provides a UART console to the device for debugging.
```

Comment:
I believe that `rpiboot` is _also_ used to load `mass-storage-gadget` onto the CM4, so perhaps this should be phrased something like:
"`mass-storage-gadget` writes devices faster than the firmware-based `rpiboot` mechanism, and also provides a UART console to the device for debugging."
?